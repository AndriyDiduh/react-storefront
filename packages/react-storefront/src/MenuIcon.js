/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { PureComponent, Fragment } from 'react'
import { observer, inject } from 'mobx-react'
import { withStyles } from '@material-ui/core'
import PropTypes from 'prop-types'
import classnames from 'classnames'

export const styles = theme => ({
  root: {
    position: 'relative',
    height: '24px',
    width: '24px',
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',

    '& .rsf-hamburger': {
      display: 'block',
      marginTop: '5px'
    },

    '& .rsf-hamburger-box': {
      width: '20px',
      height: '20px',
      display: 'block',
      position: 'relative'
    },

    '& .rsf-hamburger-inner': {
      display: 'block',
      top: '50%',
      marginTop: '-4px'
    },

    '& .rsf-hamburger-inner, & .rsf-hamburger-inner::before, & .rsf-hamburger-inner::after': {
      width: '20px',
      height: '2px',
      backgroundColor: theme.palette.text.secondary,
      borderRadius: '4px',
      position: 'absolute',
      transitionProperty: 'transform',
      transitionDuration: theme.transitions.duration.standard,
      transitionTimingFunction: 'ease'
    },

    '& .rsf-hamburger-inner::before, & .rsf-hamburger-inner::after': {
      content: '""',
      display: 'block'
    },

    '& .rsf-hamburger-inner::before': {
      top: '-10px'
    },

    '& .rsf-hamburger-inner::after': {
      bottom: '-10px'
    },

    '& .rsf-hamburger .rsf-hamburger-inner': {
      top: '5px'
    },

    '& .rsf-hamburger .rsf-hamburger-inner::before': {
      top: '5px',
      transitionProperty: 'transform, opacity',
      transitionTimingFunction: 'ease',
      transitionDuration: theme.transitions.duration.standard
    },

    '& .rsf-hamburger .rsf-hamburger-inner::after': {
      top: '10px'
    }
  },

  active: {
    '& .rsf-hamburger .rsf-hamburger-inner': {
      transform: 'translate3d(0, 5px, 0) rotate(45deg)'
    },

    '& .rsf-hamburger .rsf-hamburger-inner::before': {
      transform: 'rotate(-45deg) translate3d(-5.71429px, -5px, 0)',
      opacity: 0
    },

    '& .rsf-hamburger .rsf-hamburger-inner::after': {
      transform: 'translate3d(0, -10px, 0) rotate(-90deg)'
    }
  },

  withLabel: {
    '& .rsf-hamburger': {
      marginTop: '0'
    }
  },

  hidden: {
    display: 'none'
  },

  visible: {
    display: 'block'
  },

  label: {
    fontSize: '9px',
    lineHeight: '9px',
    fontWeight: 'bold',
    letterSpacing: '0px',
    marginTop: '-3px',
    color: theme.palette.text.secondary
  }
})

/**
 * A menu icon that animates transitions between open and closed states.
 */
@inject('app')
@observer
@withStyles(styles, { name: 'RSFMenuIcon' })
export default class MenuIcon extends PureComponent {
  static propTypes = {
    /**
     * Set to true when the menu is open, otherwise false
     */
    open: PropTypes.bool,

    /**
     * Set to `false` to hide the text "menu" underneath the button when the menu is closed.
     */
    label: PropTypes.bool
  }

  static defaultProps = {
    open: false,
    label: true
  }

  renderIcon = (open, label) => {
    const { classes } = this.props
    return (
      <div
        className={classnames({
          [classes.root]: true,
          [classes.active]: open,
          [classes.withLabel]: label
        })}
      >
        <div className="rsf-hamburger">
          <span className="rsf-hamburger-box">
            <span className="rsf-hamburger-inner" />
          </span>
        </div>
        {label && <div className={classes.label}>menu</div>}
      </div>
    )
  }

  render() {
    const {
      open,
      classes,
      label,
      app: { amp }
    } = this.props

    if (amp) {
      return (
        <Fragment>
          <div amp-bind={`class=>menuOpen == false ? '${classes.visible}' : '${classes.hidden}'`}>
            {this.renderIcon(false)}
          </div>
          <div
            className={classes.hidden}
            amp-bind={`class=>menuOpen == true ? '${classes.visible}' : '${classes.hidden}'`}
          >
            {this.renderIcon(true)}
          </div>
        </Fragment>
      )
    }

    return this.renderIcon(open, label)
  }
}
