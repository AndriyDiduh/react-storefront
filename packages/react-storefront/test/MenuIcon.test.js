import React from 'react'
import { mount } from 'enzyme'
// import ExpandIcon from '@material-ui/icons/Add'
// import CollapseIcon from '@material-ui/icons/Remove'
import TestProvider from './TestProvider'
import MenuIcon from '../src/MenuIcon'

describe('MenuIcon', () => {
  it('should show the open icon when closed', () => {
    expect(
      mount(
        <TestProvider app={{}}>
          <MenuIcon />
        </TestProvider>
      )
    ).toMatchSnapshot()
  })

  it('should show the close icon when opened', () => {
    expect(
      mount(
        <TestProvider app={{}}>
          <MenuIcon open />
        </TestProvider>
      )
    ).toMatchSnapshot()
  })

  it('should render a label', () => {
    expect(
      mount(
        <TestProvider app={{}}>
          <MenuIcon label />
        </TestProvider>
      )
    ).toMatchSnapshot()
  })

  it('should render AMP version of closed menu icon', () => {
    expect(
      mount(
        <TestProvider app={{ amp: true }}>
          <MenuIcon />
        </TestProvider>
      )
    ).toMatchSnapshot()
  })

  it('should render AMP version of open menu icon', () => {
    expect(
      mount(
        <TestProvider app={{ amp: true }}>
          <MenuIcon open />
        </TestProvider>
      )
    ).toMatchSnapshot()
  })
})
