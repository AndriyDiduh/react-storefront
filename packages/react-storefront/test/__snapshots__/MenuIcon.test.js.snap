// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuIcon should render AMP version of closed menu icon 1`] = `
<TestProvider app={{...}}>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <inject-WithStyles(MenuIcon)-with-app>
        <WithStyles(MenuIcon) app={{...}}>
          <MenuIcon app={{...}} classes={{...}} open={false} label={true}>
            <div amp-bind="class=>menuOpen == false ? \\'RSFMenuIcon-visible-23\\' : \\'RSFMenuIcon-hidden-22\\'">
              <div className="RSFMenuIcon-root-19">
                <div className="rsf-hamburger">
                  <span className="rsf-hamburger-box">
                    <span className="rsf-hamburger-inner" />
                  </span>
                </div>
              </div>
            </div>
            <div className="RSFMenuIcon-hidden-22" amp-bind="class=>menuOpen == true ? \\'RSFMenuIcon-visible-23\\' : \\'RSFMenuIcon-hidden-22\\'">
              <div className="RSFMenuIcon-root-19 RSFMenuIcon-active-20">
                <div className="rsf-hamburger">
                  <span className="rsf-hamburger-box">
                    <span className="rsf-hamburger-inner" />
                  </span>
                </div>
              </div>
            </div>
          </MenuIcon>
        </WithStyles(MenuIcon)>
      </inject-WithStyles(MenuIcon)-with-app>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;

exports[`MenuIcon should render AMP version of open menu icon 1`] = `
<TestProvider app={{...}}>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <inject-WithStyles(MenuIcon)-with-app open={true}>
        <WithStyles(MenuIcon) open={true} app={{...}}>
          <MenuIcon open={true} app={{...}} classes={{...}} label={true}>
            <div amp-bind="class=>menuOpen == false ? \\'RSFMenuIcon-visible-29\\' : \\'RSFMenuIcon-hidden-28\\'">
              <div className="RSFMenuIcon-root-25">
                <div className="rsf-hamburger">
                  <span className="rsf-hamburger-box">
                    <span className="rsf-hamburger-inner" />
                  </span>
                </div>
              </div>
            </div>
            <div className="RSFMenuIcon-hidden-28" amp-bind="class=>menuOpen == true ? \\'RSFMenuIcon-visible-29\\' : \\'RSFMenuIcon-hidden-28\\'">
              <div className="RSFMenuIcon-root-25 RSFMenuIcon-active-26">
                <div className="rsf-hamburger">
                  <span className="rsf-hamburger-box">
                    <span className="rsf-hamburger-inner" />
                  </span>
                </div>
              </div>
            </div>
          </MenuIcon>
        </WithStyles(MenuIcon)>
      </inject-WithStyles(MenuIcon)-with-app>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;

exports[`MenuIcon should render a label 1`] = `
<TestProvider app={{...}}>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <inject-WithStyles(MenuIcon)-with-app label={true}>
        <WithStyles(MenuIcon) label={true} app={{...}}>
          <MenuIcon label={true} app={{...}} classes={{...}} open={false}>
            <div className="RSFMenuIcon-root-13 RSFMenuIcon-withLabel-15">
              <div className="rsf-hamburger">
                <span className="rsf-hamburger-box">
                  <span className="rsf-hamburger-inner" />
                </span>
              </div>
              <div className="RSFMenuIcon-label-18">
                menu
              </div>
            </div>
          </MenuIcon>
        </WithStyles(MenuIcon)>
      </inject-WithStyles(MenuIcon)-with-app>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;

exports[`MenuIcon should show the close icon when opened 1`] = `
<TestProvider app={{...}}>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <inject-WithStyles(MenuIcon)-with-app open={true}>
        <WithStyles(MenuIcon) open={true} app={{...}}>
          <MenuIcon open={true} app={{...}} classes={{...}} label={true}>
            <div className="RSFMenuIcon-root-7 RSFMenuIcon-active-8 RSFMenuIcon-withLabel-9">
              <div className="rsf-hamburger">
                <span className="rsf-hamburger-box">
                  <span className="rsf-hamburger-inner" />
                </span>
              </div>
              <div className="RSFMenuIcon-label-12">
                menu
              </div>
            </div>
          </MenuIcon>
        </WithStyles(MenuIcon)>
      </inject-WithStyles(MenuIcon)-with-app>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;

exports[`MenuIcon should show the open icon when closed 1`] = `
<TestProvider app={{...}}>
  <Provider router={{...}} app={{...}} history={{...}}>
    <MuiThemeProviderOld theme={{...}}>
      <inject-WithStyles(MenuIcon)-with-app>
        <WithStyles(MenuIcon) app={{...}}>
          <MenuIcon app={{...}} classes={{...}} open={false} label={true}>
            <div className="RSFMenuIcon-root-1 RSFMenuIcon-withLabel-3">
              <div className="rsf-hamburger">
                <span className="rsf-hamburger-box">
                  <span className="rsf-hamburger-inner" />
                </span>
              </div>
              <div className="RSFMenuIcon-label-6">
                menu
              </div>
            </div>
          </MenuIcon>
        </WithStyles(MenuIcon)>
      </inject-WithStyles(MenuIcon)-with-app>
    </MuiThemeProviderOld>
  </Provider>
</TestProvider>
`;
